src/scraping/scrape_activities.py

```diff
@@ -0,0 +1,158 @@
+
+# Possible headers signalling the end of the article
+END_MARKERS = [
+    "In case you missed it",
+    "Find another activity",
+]
+
+# Boilerplate text to be removed
+SCRAP_TEXT = [
+    "TwitterFacebookMessengerWhatsAppShareShare this withTwitterFacebookMessengerWhatsAppCopy linkRead more about sharing",
+    "previousnext",
+]
+
+
+def pad_element(element: bs4.element.Tag) -> str:
+    """Pad an element's content with spaces for certain tags (a recursive function)
+
+    Args:
+        element (bs4.element.Tag): HTML element
+
+    Returns:
+        str: Text with spaces
+    """
+    if isinstance(element, str):
+        return element
+    elif element.name in ["p", "h1", "h2", "h3", "h4", "h5", "h6"]:
+        return " " + " ".join(pad_element(child) for child in element.children) + " "
+    else:
+        return "".join(pad_element(child) for child in element.children)
+
+
+# Function to scrape the web page
+def web_scraper(url: str) -> str:
+    """Scrape a web page and return the content
+
+    Args:
+        url
+            URL of the web page
+
+    Returns:
+        Dataframe with URL and content
+    """
+    # Fetch webpage
+    response = requests.get(
+        url,
+        timeout=10,
+        headers={
+            "User-Agent": "Data collection for the purpose of research. For questions, reach out to karlis.kanders@nesta.org.uk"
+        },
+    )
+
+    # Parse with BeautifulSoup
+    soup = BeautifulSoup(response.text, "html.parser")
+
+    # Remove all <style>...</style> tags
+    for style_tag in soup.find_all("style"):
+        style_tag.extract()
+
+    # Find the headline
+    headline = soup.find("h1", class_="blocks-article__headline")
+
+    # Fetch all the content starting from headline
+    content = []
+    curr_element = headline.find_next_sibling()
+    while curr_element is not None:
+        content.append(curr_element)
+        curr_element = curr_element.find_next_sibling()
+
+    # Clean the text from HTML tags and add spaces
+    clean_content = "".join(pad_element(element) for element in content)
+
+    # Remove the end of the webpage
+    if any(marker in clean_content for marker in END_MARKERS):
+        for marker in END_MARKERS:
+            clean_content = clean_content.split(marker)[0]
+
+    # Remove the boilerplate text
+    for text in SCRAP_TEXT:
+        clean_content = clean_content.replace(text, "")
+
+    return clean_content.strip()
+
+
+if __name__ == "__main__":
+    urls_df = pd.read_csv(URL_PATH)
+
+    # Fetch the already scraped urls
+    if SCRAPED_PATH.exists():
+        scraped_urls = pd.read_csv(SCRAPED_PATH, names=["URL", "text"]).URL.to_list()
+    else:
+        scraped_urls = []
+
+    # Scrape the urls
+    with open(SCRAPED_PATH, "a") as f:
+        writer = csv.writer(f)
+        for _, row in tqdm(urls_df.iterrows(), total=len(urls_df)):
+            # Check if the url has already been scraped
+            if row.URL in scraped_urls:
```

Comment:
You could remove the `if` statement by filtering the scraped URLs earlier in the script with:

```python
urls_df[~urls_df["URL"].isin(scraped_urls)]
```